[13.0]]ADD] Add account_invoicing_mode

TDu opened the pull request:
Add three modules helping on automatically invoicing customers.

The base module `account_invoice_base_invoicing_mode` does not actually
do anything but adds a selection field on partner to allow assigning an
invoicing mode to a customer.

The two other modules add specific invoicing mode.

The module `account_invoice_mode_monthly` creates monthly invoices for
customer on a specific day (configruation is in Accounting Settings)

The module `account_invoice_mode_at_shipping` creates invoices on the
shipping of the goods.

Those modules use queue_job to generate and validate invoices.

TDu committed:
[13.0]]ADD] Add acconut_invoicing_mode

Add three modules helping on automatically invoicing customers.

The base module `account_invoice_base_invoicing_mode` does not actually
do anything but adds a selection field on partner to allow assigning an
invoicing mode to a customer. And a checkbox to choose regrouping
invoices.

The two other modules add specific invoicing mode.

The module `account_invoice_mode_monthly` creates monthly invoices for
customer on a specific day (configuration is in Accounting Settings)

The module `account_invoice_mode_at_shipping` creates invoices on the
shipping of the goods.

Those modules use queue_job to generate and validate invoices.

jgrandguillaume reviewed (approved):
Functional test, looks good. Thx for the job

jbaudoux commented on account_invoice_mode_monthly/data/ir_cron.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="numbercall">-1</field>
+        <field eval="False" name="doall" />
+        <field name="model_id" ref="model_sale_order" />
+        <field name="code">model.cron_generate_montly_invoices()</field>
```
```suggestion
        <field name="code">model.cron_generate_monthly_invoices()</field>
```

TDu replied:
Fixed

jbaudoux commented on account_invoice_mode_monthly/models/sale_order.py:
```diff
@@ -0,0 +1,100 @@
+    _inherit = "sale.order"
+
+    @api.model
+    def cron_generate_montly_invoices(self):
```
```suggestion
    def cron_generate_monthly_invoices(self):
```

TDu replied:
fixed

jbaudoux commented on account_invoice_mode_monthly/models/sale_order.py:
```diff
@@ -0,0 +1,100 @@
+            ]
+        )
+        # By default grouped by partner/currency. Refund are not generated
+        invoices = sales._create_invoices(grouped=partner.one_invoice_per_order)
```
Do not group sales having different Payment Term or Payment Mode
It would have been nice to be able to set a specific SO as "one_invoice_per_order"

TDu replied:
@jbaudoux I changed the machinery so sales orders with different payment terms are not grouped into the same invoice.
This is done before the creation of the jobs (by the cron). And can be easily extended.
Also for the payment mode this part of an other OCA module so will not be done here

jbaudoux commented on account_invoice_mode_monthly/models/sale_order.py:
```diff
@@ -0,0 +1,100 @@
+        )
+        # By default grouped by partner/currency. Refund are not generated
+        invoices = sales._create_invoices(grouped=partner.one_invoice_per_order)
+        for invoice in invoices:
```
Please also add a block for refunds where order_line.qty_to_invoice < 0 and call _create_invoice with final=True

TDu replied:
I have changed the domain above with `qty_to_invoice` != 0 and set the final to True. That should do the trick

jbaudoux reviewed (changes requested):
Some remarks (typo, refunds)

jbaudoux commented on account_invoice_base_invoicing_mode/models/sale_order.py:
```diff
@@ -0,0 +1,15 @@
+
+    invoicing_mode = fields.Selection(related="partner_invoice_id.invoicing_mode")
+
+    # def _create_invoices(self, grouped=False, final=False):
```
To remove

jbaudoux commented on account_invoice_base_invoicing_mode/models/sale_order.py:
```diff
@@ -0,0 +1,15 @@
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    invoicing_mode = fields.Selection(related="partner_invoice_id.invoicing_mode")
```
Remove or set readonly

TDu replied:
Set readonly

jbaudoux commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,32 @@
+    def _invoicing_at_shipping(self):
+        self.ensure_one()
+        sale_order_ids = self._get_sales_order_to_invoice()
+        invoices = sale_order_ids._create_invoices()
```
Can you also take care of passing grouped=partner.one_invoice_per_order in the same logic as you did above ?

TDu replied:
Done

jbaudoux commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,32 @@
+        return invoices
+
+    def _get_sales_order_to_invoice(self):
+        return self.sale_id
```
self.mapped("move_lines.group_id.sale_id")

TDu replied:
Done

jbaudoux reviewed (commented):
Some additional remarks to be compatible with the module https://github.com/OCA/stock-logistics-workflow/tree/13.0/stock_picking_group_by_partner_by_carrier

mmequignon commented on account_invoice_base_invoicing_mode/tests/test_base_invoicing_mode.py:
```diff
@@ -0,0 +1,11 @@
+from odoo.tests.common import SavepointCase
+
+
+class TestBaseInvoicingMode(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
No tests here, is it normal?

TDu replied:
There is no test because it is the base module that does not anything without a specific mode.
Maybe could be removed ?

TDu replied:
It has been removed

mmequignon commented on account_invoice_mode_monthly/models/sale_order.py:
```diff
@@ -0,0 +1,100 @@
+        return invoices
+
+    @api.model
+    def _company_monthly_invoicing_today(self):
+        """Get company ids for which today is monthly invoicing day."""
+        today = datetime.now()
+        first_day_this_month = today.replace(
+            day=1, hour=0, minute=0, second=0, microsecond=0
+        )
+        first_day_last_month = first_day_this_month - relativedelta(months=1)
+        last_day_of_this_month = (today + relativedelta(day=31)).day
+        # Last month still not executed, it needs to be done
+        domain_last_month = [
+            ("invoicing_mode_monthly_last_execution", "<", first_day_last_month),
+        ]
+        # Invoicing day is today or in the past and invoicing not yet done
+        domain_this_month = [
+            "|",
+            ("invoicing_mode_monthly_last_execution", "<", first_day_this_month),
+            ("invoicing_mode_monthly_last_execution", "=", False),
+            ("invoicing_mode_monthly_day_todo", "<=", today.day),
+        ]
+        # Make sure non exisiting days are done at the end of the month
+        domain_last_day_of_month = [
+            "|",
+            ("invoicing_mode_monthly_last_execution", "<", first_day_this_month),
+            ("invoicing_mode_monthly_last_execution", "=", False),
+            ("invoicing_mode_monthly_day_todo", ">", today.day),
+        ]
+        if today.day == last_day_of_this_month:
+            domain = OR(
+                [domain_last_month, domain_this_month, domain_last_day_of_month]
+            )
+        else:
+            domain = OR([domain_last_month, domain_this_month])
+
+        return self.env["res.company"].search(domain)
```
I feel like that method is kind of complicated for what it does.
Would it be easier to compute the next invoicing day at the end of the invoicing, and return companies where `("next_invoice_day", "=", "fields.date.today()")` ?

TDu replied:
The complication was added because you never know when a cron is run. A daily cron could just not have run for multiple days.

mmequignon replied:
Yes, then for any company with `("next_invoice_day", "<=", "field.date.today()")`, that method should be run.
You just have to change the operator. Once that method is run on a record, you can update the date on it.
Except for the special case you mentioned, you shouldn't have some `next_invoice_day` inferior to today, am I right ?

mmequignon replied:
I probably don't have the full picture of it, though.

mmequignon replied:
Also, it may be complicated to compute the next month given a day.

simahawk replied:
@TDu ?

TDu replied:
I think this has been answered by this comment https://github.com/OCA/account-invoicing/pull/755#discussion_r625023369

pedrobaeza commented:
I think this module is not correctly focused by several reasons:

- The aim is to invoice sales order, so the name and the repo are not good.
- It's overlapping `sale_automatic_workflow` flow.

IMO, you should extend that module for adding the monthly/dayly mechanism, but using the same base methods.

TDu committed:
[13.0][IMP] acconut_invoicing_mode group by terms

Sales order with different payment term are not grouped into the
same invoice anymore.
This grouping can be easily changed by overriding the coresponding
method.

During this change the parameter of the job generating the invoices
has changes and is now the sale_order_ids to process.

Commit:
Invoice sales based on moves

If we group several sales in the same procurement group, we'll invoice
only one sale order. Tracing back to the sales through the stock moves
ensure we invoice all of them.

Commit:
Ensure _get_related_invoices include all the related invoices

Through the stock moves

simahawk commented on account_invoice_base_invoicing_mode/models/sale_order.py:
```diff
@@ -0,0 +1,12 @@
+    _inherit = "sale.order"
+
+    invoicing_mode = fields.Selection(
+        related="partner_invoice_id.invoicing_mode", readonly=True
```
```suggestion
        related="partner_invoice_id.invoicing_mode"
```

TDu replied:
Well, I just noticed that it was first declared without the readonly, then it was added later on after a review.
I guess there is some good reason to have it or not !?

simahawk replied:
readonly is true by default on v13 so you must specify it when you don't want it.

TDu replied:
Oh yes you are correct so it was unnecessary code :+1:

simahawk commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,35 @@
+
+    def action_done(self):
+        res = super().action_done()
+        picking_to_invoice = self.filtered(
+            lambda r: r.sale_id.partner_invoice_id.invoicing_mode == "at_shipping"
+            and r.picking_type_code == "outgoing"
+        )
+        for picking in picking_to_invoice:
+            picking.with_delay()._invoicing_at_shipping()
+        return res
```
Avoid looping twice and make it more readable and customizable:
```suggestion
        for picking in picking_to_invoice:
            if picking._invoice_at_shipping():
                picking.with_delay()._invoicing_at_shipping()
        return res
        
    def _invoice_at_shipping(self):
        """Check if current record must be invoiced at shipping."""
        return (
            self.picking_type_code == "outgoing" and
            self.sale_id.partner_invoice_id.invoicing_mode == "at_shipping"
        )
```

TDu replied:
This has been changed

simahawk commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,35 @@
+    @job(default_channel="root.invoice_at_shipping")
+    def _invoicing_at_shipping(self):
+        self.ensure_one()
+        sale_order_ids = self._get_sales_order_to_invoice()
+        partner = sale_order_ids.partner_invoice_id
+        invoices = sale_order_ids._create_invoices(
+            grouped=partner.one_invoice_per_order
+        )
```
```suggestion
        sale_orders = self._get_sales_order_to_invoice()
        partner = sale_orders.partner_invoice_id
        invoices = sale_orders._create_invoices(
            grouped=partner.one_invoice_per_order
        )
```

TDu replied:
This had been resolved

simahawk commented on account_invoice_mode_at_shipping/tests/test_invoice_mode_at_shipping.py:
```diff
@@ -0,0 +1,77 @@
+class TestInvoiceModeAtShipping(SavepointCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
```
```suggestion
        super().setUpClass()
        cls.env = cls.env(context=dict(cls.env.context, tracking_disable=True))
```
you could move this and the creation of sale order in the base class (which is not used here).
Otherwise, remove the base class.

TDu replied:
I removed the base class

simahawk commented on account_invoice_mode_at_shipping/tests/test_invoice_mode_at_shipping.py:
```diff
@@ -0,0 +1,77 @@
+        self.assertEqual(self.so1.invoice_ids.state, "posted")
+
+    def test_invoice_not_created_at_shipping(self):
+        """Check that an invoice is created when goods are shipped."""
```
```suggestion
        """Check that an invoice is not created when goods are shipped."""
```

simahawk commented on account_invoice_mode_monthly/models/sale_order.py:
```diff
@@ -0,0 +1,106 @@
+        for saleorder_group in saleorder_groups:
+            saleorder_ids = self.search(saleorder_group["__domain"]).ids
+            self.with_delay()._generate_invoices_by_partner(saleorder_ids)
+        companies.write({"invoicing_mode_monthly_last_execution": datetime.now()})
```
why don't you compute and store the next invoice day instead? Your search query will be much more simple I think, no?

TDu replied:
The query may look complicated but it is covered by tests.
I am not convince that storing the next execution date would make things simpler. One of the reason is the day of invoicing in the month can be set in the configuration.
So now when changing that day the next invoicing day as too be computed but depending if the invoicing day as been done for the current month or not the computation would be different. (This is only one complication I can think of now)

TDu committed:
[13.0][FIX] Add acconut_invoicing_mode

Add few changes from reviews.
Use sudo to post the invoice.

i-vyshnevska commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,39 @@
+    @job(default_channel="root.invoice_at_shipping")
+    def _invoicing_at_shipping(self):
+        self.ensure_one()
+        sales_order = self._get_sales_order_to_invoice()
```
I think you may need to filter sale order here if there are invocable lines, otherwise, it will raise an error 
https://github.com/odoo/odoo/blob/be6b59fb3bd52894524603ee81597d8ec289583a/addons/sale/models/sale.py#L629

TDu replied:
I have added a check for it, thanks.

TDu committed:
[13.0][FIX] account_invoicing_mode_at_shipping

Split invoice creation on invoice partner settings that decides if sales order
atre being grouped in invoice or not.
The previous implementation would crash.

TDu committed:
[13.0][FIX] account_invoicing_mode_at_shipping

Filter out sales order that have nothing to invoice.
To avoid an exception at invoice createion.

TDu committed:
[13.0][IMP] account_invoicing_mode_at_shipping

Add explicit message if nothing is invoiced

jbaudoux commented on account_invoice_mode_at_shipping/models/stock_picking.py:
```diff
@@ -0,0 +1,51 @@
+        sales = self.env["sale.order"].browse()
+        # Filter out non invoicable sales order
+        for sale in self._get_sales_order_to_invoice():
+            if sale._get_invoiceable_lines():
```
@TDu This should be part of the above method `_get_sales_order_to_invoice` that has to return only orders having invoiceable lines

TDu replied:
done

TDu commented:
In previous comments there has been a suggestion by @pedrobaeza to include this into `sale_automatic_workflow` because the use case is overlapping  with it.
I don't think so myself, the monthly invoice mode is based on a daily cron, the option does not exist in `sale_automatic_workflow` and the at shipping invoicing is triggered by some specific code in the action_done of the stock_picking, which would not be easily implemented on top of `sale_automatic_workflow`.

Moreover the proposed change would also make this pr much more complex and maybe its UI element less intuitive to use.

And I personally think that this pr gives a good base to add different invoicing mode in the future.

pedrobaeza commented:
I don't remember right now my suggestion and the context, so I can't give you now an answer, but if you point me to it, I can agree or not on this.

TDu commented:
@pedrobaeza It is about this comment https://github.com/OCA/account-invoicing/pull/755#issuecomment-662846222 on the same PR.
We would like to get this merge and I felt your proposition still deserved an answer, yep that is one year ago.

sebalix reviewed (approved):
LG, a rebase to keep one commit / module is welcome

pedrobaeza commented:
OK, as you prefer then. I still encounter some common points:

- `sale_automatic_workflow` also has a cron that performs the task of invoicing.
- It contains the code to invoice.

What about the naming, being `sale_invoicing_mode` or similar?

TDu commented:
> OK, as you prefer then. I still encounter some common points:
> 
>     * `sale_automatic_workflow` also has a cron that performs the task of invoicing.
> 
>     * It contains the code to invoice.
> 
> 
> What about the naming, being `sale_invoicing_mode` or similar?

Well the base module is just a field on the partner really, so it is about invoicing first. So it could be used to invoice other things than a sale order if the business model requires that.

I would really like to keep the naming like it is.

pedrobaeza commented:
What about partner_invoicing_mode then? If not, I won't block it.

TDu commented:
> What about partner_invoicing_mode then? If not, I won't block it.

I actually like that name better, shorter and more explicit.
Maybe it could be renamed during the 14.0 migration

TDu committed:
[13.0] account-invoicing-mode add roadmap

TDu commented:
@simahawk I squashed the commit and added the roadmap as discussed

simahawk commented:
@TDu check the build pls

simahawk reviewed (approved):
@jbaudoux @mmequignon good for you?

mmequignon commented on account_invoice_base_invoicing_mode/models/account_invoice.py:
```diff
@@ -0,0 +1,15 @@
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    @job(default_channel="root.invoice_validation")
```
Isn't the @job decorator deprecated?
https://github.com/OCA/queue/pull/274

TDu replied:
Yes it is, but it still works for v13.
It will be changed on the v14 migration.

simahawk commented:
/ocabot merge patch

simahawk commented:
/ocabot merge patch

simahawk commented:
WTH?

![image](https://user-images.githubusercontent.com/347149/127810472-c3c2d4a2-2207-4754-b688-d5f899bda0aa.png)

sbidoul commented:
If you click on the link in the error message, you'll see that travis is indeed red (a transient GitHub error). It is not the same commit as the one just above for a good reason I don't remember right now.

![image](https://user-images.githubusercontent.com/692075/127863403-70c584e8-ee03-40e9-bfb3-875ec4b97b59.png)

simahawk commented:
let's try again 
/ocabot merge patch
